src/bc/model.rs

```diff
@@ -22,10 +23,184 @@ pub enum BcBody {
     ModernMsg(ModernMsg),
 }
 
+pub const MAGIC_VIDEO_INFO: &[u8] = &[0x31, 0x30, 0x30, 0x31];
+pub const MAGIC_AAC: &[u8] = &[0x30, 0x35, 0x77, 0x62];
+pub const MAGIC_ADPCM: &[u8] = &[0x30, 0x31, 0x77, 0x62];
+pub const MAGIC_IFRAME:  &[u8] = &[0x30, 0x30, 0x64, 0x63];
+pub const MAGIC_PFRAME:  &[u8] = &[0x30, 0x31, 0x64, 0x63];
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct VideoIFrame {
+    pub raw: Vec<u8>
+}
+
+impl VideoIFrame {
+    pub fn from_binary(binary: &[u8]) -> VideoIFrame {
+        VideoIFrame {
+            raw: binary.to_vec(),
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.raw.len()
+    }
+
+    pub fn as_slice(&self) -> &[u8] {
+        self.raw.as_slice()
+    }
+
+    pub fn magic(&self) -> &[u8] {
+        &self.raw[0..4]
+    }
+
+    pub fn video_type(&self) -> &[u8] {
+        &self.raw[4..8]
+    }
+
+    pub fn data_size(&self) -> usize {
+        u32::from_be_bytes(self.raw[8..12].try_into().expect("slice with incorrect length")).try_into().expect("u32 won't fit into usize")
+    }
+
+    pub fn unknowna(&self) -> &[u8] {
+        &self.raw[12..16]
+    }
+
+    pub fn timestamp(&self) -> &[u8] {
+        &self.raw[16..20]
+    }
+
+    pub fn unknownb(&self) -> &[u8] {
+        &self.raw[20..24]
+    }
+
+    pub fn clocktime(&self) -> &[u8] {
+        &self.raw[24..28]
+    }
+
+    pub fn unknownc(&self) -> &[u8] {
+        &self.raw[28..32]
+    }
+
+    pub fn video_data(&self) -> &[u8] {
+        &self.raw[32..]
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct VideoPFrame {
+    pub raw: Vec<u8>
+}
+
+impl VideoPFrame {
```

Comment:
I think you should have a single `BinaryData` struct like the following:

```rust
pub struct BinaryData {
    data: Vec<u8>,
}

impl BinaryData {
    pub fn body(&self) -> &[u8] { /* ... */ }
    pub fn kind(&self) -> BinaryDataKind { /* match on header */ }
}
```

and don't bother with parsing the header much right now.  We can add helpers to give `bc_protocol` more info if it needs it, and this way you have a single struct to deal with, with none of the duplication that IFrame, PFrame causes.

Reply:
`bc_protocol` can `match` on the `BinaryDataKind`. This separates the "control path" from the "data path", keeping the latter using a single data type.